src/trace/implementations/merge_batcher.rs

```diff
@@ -224,170 +229,466 @@ pub trait Merger: Default {
-        } else {
-            1
+    /// Behavior to dissect items of chunks in the merge batcher
+    pub trait MergerChunk : SizableContainer {
+        /// An owned time type.
+        ///
+        /// This type is provided so that users can maintain antichains of something, in order to track
+        /// the forward movement of time and extract intervals from chains of updates.
+        type TimeOwned;
+        /// The owned diff type.
+        ///
+        /// This type is provided so that users can provide an owned instance to the `push_and_add` method,
+        /// to act as a scratch space when the type is substantial and could otherwise require allocations.
+        type DiffOwned: Default;
+
+        /// Relates a borrowed time to antichains of owned times.
+        ///
+        /// If `upper` is less or equal to `time`, the method returns `true` and ensures that `frontier` reflects `time`.
+        fn time_kept(time1: &Self::Item<'_>, upper: &AntichainRef<Self::TimeOwned>, frontier: &mut Antichain<Self::TimeOwned>) -> bool;
+
+        /// Push an entry that adds together two diffs.
+        ///
+        /// This is only called when two items are deemed mergeable by the container queue.
+        /// If the two diffs added together is zero do not push anything.
+        fn push_and_add<'a>(&mut self, item1: Self::Item<'a>, item2: Self::Item<'a>, stash: &mut Self::DiffOwned);
+
+        /// Account the allocations behind the chunk.
+        // TODO: Find a more universal home for this: `Container`?
+        fn account(&self) -> (usize, usize, usize, usize) {
+            let (size, capacity, allocations) = (0, 0, 0);
+            (self.len(), size, capacity, allocations)
         }
     }
+    
+    /// A merger for arbitrary containers.
+    ///
+    /// `MC` is a [`Container`] that implements [`MergerChunk`].
+    /// `CQ` is a [`ContainerQueue`] supporting `MC`.
+    pub struct ContainerMerger<MC, CQ> {
+        _marker: PhantomData<(MC, CQ)>,
+    }
 
-    /// Helper to get pre-sized vector from the stash.
-    #[inline]
-    fn empty(&self, stash: &mut Vec<Vec<T>>) -> Vec<T> {
-        stash.pop().unwrap_or_else(|| Vec::with_capacity(self.chunk_capacity()))
+    impl<MC, CQ> Default for ContainerMerger<MC, CQ> {
+        fn default() -> Self {
+            Self { _marker: PhantomData, }
+        }
     }
 
-    /// Helper to return a chunk to the stash.
-    #[inline]
-    fn recycle(&self, mut chunk: Vec<T>, stash: &mut Vec<Vec<T>>) {
-        // TODO: Should we limit the size of `stash`?
-        if chunk.capacity() == self.chunk_capacity() {
-            chunk.clear();
+    impl<MC: MergerChunk, CQ> ContainerMerger<MC, CQ> {
+        /// Helper to get pre-sized vector from the stash.
+        #[inline]
+        fn empty(&self, stash: &mut Vec<MC>) -> MC {
+            stash.pop().unwrap_or_else(|| {
+                let mut container = MC::default();
+                container.ensure_capacity(&mut None);
```

Comment:
At the moment, we have a different capacity for columnation on the wire vs. in the merge batcher, and if I remember correctly, the reason for this is that larger columnation objects play nicer with lgalloc because their spilled information is potentially larger. I just want to point this out so we don't forget, but at the same time it's a good moment to revisit this behavior.

Reply:
We discussed, and the conclusion for now is that this is ok to merge, as one can work around it by wrapping `MC` appropriately, with something that has new opinions about what `at_capacity()` and `ensure_capacity()` should do.